jax/lax/lax.py

```diff
@@ -3952,7 +3952,131 @@ def _scatter_dtype_rule(operand, scatter_indices, updates, **kwargs):
   _check_same_dtypes("scatter", False, operand.dtype, updates.dtype)
   return dtypes.canonicalize_dtype(operand.dtype)
 
-def _scatter_shape_rule(operand, scatter_indices, updates, **kwargs):
+def _scatter_shape_rule(operand, scatter_indices, updates, *, update_jaxpr,
+                        update_consts, dimension_numbers, indices_are_sorted,
+                        unique_indices):
+  """Validates the well-formedness of the ``dimension_numbers`` argument to
+  Scatter.
+
+  The code implements the checks based on the detailed operation semantics of
+  XLA's `Scatter <https://www.tensorflow.org/xla/operation_semantics#scatter>`_
+  operator and following the outline of the implementation of
+  ShapeInference::InferScatterShape in TensorFlow.
+  """
+  rank = lambda arr: len(arr.shape)
+
+  def _is_sorted(dims, name):
+    for i in range(1, len(dims)):
+      if dims[i] < dims[i - 1]:
+        raise TypeError(f"{name} in scatter op must be sorted; got {dims}")
```

Comment:
I am wondering if it makes sense to merge _is_sorted and _sorted_dims_in_range, and check for each dim that it is within dims[i - 1] and rank?

Reply:
My rationale for doing it is the same as what I said [here](https://github.com/google/jax/pull/4144#discussion_r478229284). If we decide not to mirror the structure of the XLA shape inference too much though, then I think it indeed makes sense.